### PR TITLE
Async/Await parse

### LINKED ISSUE #818: Implement `async` / `await`
**ECMASCript feature**
We currently have no support for `async` / `await` in Boa. These two keywords are used to create "promises" that will be executed asynchronously, without blocking the execution of the code that follows.

There are two types of `async` constructs:
 - Async Function Definitions: https://tc39.es/ecma262/#sec-async-function-definitions
 - Async Arrow Function Definitions: https://tc39.es/ecma262/#sec-async-arrow-function-definitions

Implementing this has to be done in two phases. First we need to be able to parse these `async` constructs, and for that, we will need new AST nodes and new parsers, that will be called when finding the `async` keyword.

After the parsing is done, we need to implement the execution. This will be a bit tricky, since we currently just execute the nodes one by one. This means we need to follow the spec in the "Runtime Semantics" for these two constructs, and implement the needed methods.

This might require further talk in #367.

### PR DESCRIPTION
**Pull request by Lan2u**
This Pull Request addresses the parse side of #818 .

It changes the following:

- (Mainly) add support for parsing of async function decl, async function expression and await expression.
- (Side-effect) tidy up / moving of non-async function parsing into own folders + some additional tests

The run methods return undefined (rather than an unimplemented! panic) while waiting for an execution implementation.

### CODE REVIEW

File: boa/src/syntax/ast/node/await_expr/mod.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+use serde::{Deserialize, Serialize};
+
+///
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]:
+/// [mdn]:
```
Comment (HalidOdat) [inline boa/src/syntax/ast/node/await_expr/mod.rs:18]: Doc fix.

File: boa/src/syntax/ast/node/declaration/async_function_decl/mod.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        self.body.display(f, indentation + 1)?;
+
+        writeln!(f, "}}")
+    }
+}
+
+impl Executable for AsyncFunctionDecl {
+    fn run(&self, _: &mut Context) -> Result<Value> {
+        let _timer = BoaProfiler::global().start_event("AsyncFunctionDecl", "exec");
+        // unimplemented!("Execute AsyncFunctionDecl");
```
Comment (HalidOdat) [inline boa/src/syntax/ast/node/declaration/async_function_decl/mod.rs:84]:
```suggestion
        // TODO: Implement AsyncFunctionDecl
```
Better replace this with a todo

File: boa/src/syntax/ast/node/await_expr/mod.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone, Debug, Trace, Finalize, PartialEq)]
+pub struct AwaitExpr {
+    expr: Box<Node>,
+}
+
+impl Executable for AwaitExpr {
+    fn run(&self, _: &mut Context) -> Result<Value> {
+        let _timer = BoaProfiler::global().start_event("AwaitExpression", "exec");
+        // unimplemented!("Await expression execution");
```
Comment (HalidOdat) [inline boa/src/syntax/ast/node/await_expr/mod.rs:28]:
```suggestion
        // TODO: Implement AwaitExpr
```

File: boa/src/syntax/ast/node/declaration/async_function_expr/mod.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        self.body.display(f, indentation + 1)?;
+
+        writeln!(f, "}}")
+    }
+}
+
+impl Executable for AsyncFunctionExpr {
+    fn run(&self, _: &mut Context) -> Result<Value> {
+        // unimplemented!("Execute AsyncFunctionExpr");
```
Comment (HalidOdat) [inline boa/src/syntax/ast/node/declaration/async_function_expr/mod.rs:80]:
```suggestion
        // TODO: Implement AsyncFunctionExpr
```

File: boa/src/syntax/parser/statement/declaration/mod.rs [hunk 5]
```diff
@@ -63,7 +63,7 @@ where
         let tok = cursor.peek(0)?.ok_or(ParseError::AbruptEnd)?;
 
         match tok.kind() {
-            TokenKind::Keyword(Keyword::Function) => {
+            TokenKind::Keyword(Keyword::Function) | TokenKind::Keyword(Keyword::Async) => {
```
Comment (HalidOdat) [inline boa/src/syntax/parser/statement/declaration/mod.rs:66]:
```suggestion
            TokenKind::Keyword(Keyword::Function | Keyword::Async) => {
```

File: boa/src/syntax/parser/statement/mod.rs [hunk 6]
```diff
@@ -386,7 +390,7 @@ where
         let tok = cursor.peek(0)?.ok_or(ParseError::AbruptEnd)?;
 
         match *tok.kind() {
-            TokenKind::Keyword(Keyword::Function) => {
+            TokenKind::Keyword(Keyword::Function) | TokenKind::Keyword(Keyword::Async) => {
```
Comment (HalidOdat) [inline boa/src/syntax/parser/statement/mod.rs:393]:
```suggestion
            TokenKind::Keyword(Keyword::Function | Keyword::Async) => {
```

File: boa/src/syntax/parser/expression/await_expr.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
+//! Await expression parsing.
+//!
+//! More information:
+//!  - [MDN documentation][mdn]
+//!  - [ECMAScript specification][spec]
+//!
+//! [mdn]:
+//! [spec]:
```
Comment (HalidOdat) [inline boa/src/syntax/parser/expression/await_expr.rs:8]: missing doc links

File: boa/src/syntax/parser/statement/declaration/mod.rs (hunk 5 above)
Comment (Lan2u) [reply boa/src/syntax/parser/statement/declaration/mod.rs]: Ah! it seems I'm on an outdated rust version

Comment (Lan2u) [reply boa/src/syntax/parser/statement/declaration/mod.rs]: Actually, it seems this is still experimental as of 1.47.0?

Comment (HalidOdat) [reply boa/src/syntax/parser/statement/declaration/mod.rs:66]: > Actually, it seems this is still experimental as of 1.47.0?

ah sorry about that, I thought it was stable

**Review by RageKnify — APPROVED** [review]

File: boa/src/syntax/parser/statement/declaration/hoistable/async_function_decl/mod.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+impl<R> TokenParser<R> for AsyncFunctionDeclaration
+where
+    R: Read,
+{
+    type Output = AsyncFunctionDecl;
+
+    fn parse(self, cursor: &mut Cursor<R>) -> Result<Self::Output, ParseError> {
+        cursor.expect(Keyword::Async, "async function declaration")?;
+        cursor.expect_no_skip_lineterminator(Keyword::Function, "async function declaration")?;
```
Comment (Razican) [inline boa/src/syntax/parser/statement/declaration/hoistable/async_function_decl/mod.rs:53]: I would not use a new function here either, but maybe it makes sense to add a `context` string to `peek_expect_no_lineterminator()` in order to improve diagnostics.

```suggestion
        cursor.peek_expect_no_lineterminator(0)?;
        cursor.expect(Keyword::Function, "async function declaration")?;
```

File: boa/src/syntax/parser/statement/declaration/hoistable/async_function_decl/mod.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        let tok = cursor.peek(0)?;
+
+        let name = if let Some(token) = tok {
+            match token.kind() {
+                TokenKind::Punctuator(Punctuator::OpenParen) => {
+                    if !self.is_default.0 {
+                        return Err(ParseError::unexpected(
+                            token.clone(),
+                            "Unexpected missing identifier for async function decl",
```
Comment (Razican) [inline boa/src/syntax/parser/statement/declaration/hoistable/async_function_decl/mod.rs:62]: For an improved message on display:
```suggestion
                            " in async function declaration",
```

File: boa/src/syntax/parser/statement/declaration/hoistable/tests.rs [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (Razican) [inline boa/src/syntax/parser/statement/declaration/hoistable/tests.rs:1]: Why is this empty / exists?

File: boa/src/syntax/parser/statement/declaration/tests.rs [hunk 11]
```diff
@@ -169,26 +166,3 @@ fn multiple_const_declaration() {
...
-    check_parser(
-        "function yield() {}",
-        vec![FunctionDecl::new(Box::from("yield"), vec![], vec![]).into()],
-    );
-
-    check_parser(
-        "function await() {}",
-        vec![FunctionDecl::new(Box::from("await"), vec![], vec![]).into()],
-    );
-}
```
Comment (Razican) [inline boa/src/syntax/parser/statement/declaration/tests.rs:194]: Why were these removed? Were they moved elsewhere?

**Review by Razican — COMMENTED** [review]
I finally got a bit of time to review this code. It's looking very, very good. Check my comments and let's see if it can be improved!

File: boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                TokenKind::Punctuator(Punctuator::OpenParen) => None,
+                _ => Some(BindingIdentifier::new(self.allow_yield, true).parse(cursor)?),
+            }
+        } else {
+            return Err(ParseError::AbruptEnd);
+        };
+
+        cursor.expect(Punctuator::OpenParen, "function expression")?;
+
+        let params = FormalParameters::new(!self.allow_yield.0, true).parse(cursor)?;
```
[the remark this answers is not in the document]
Comment (Lan2u) [reply boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs:67]: Doesn't ~yield mean NOT yield rather than false?

File: boa/src/syntax/parser/statement/declaration/tests.rs (hunk 11 above)
Comment (Lan2u) [reply boa/src/syntax/parser/statement/declaration/tests.rs:194]: Yeah they were split into function_decl tests

File: boa/src/syntax/parser/statement/declaration/hoistable/tests.rs (hunk 10 above)
Comment (Lan2u) [reply boa/src/syntax/parser/statement/declaration/hoistable/tests.rs:1]: I think this highlights we don't actually have any tests specifically targeting hoistable statement decl except those for function decl / async function decl

File: boa/src/syntax/parser/statement/declaration/hoistable/async_function_decl/mod.rs (hunk 8 above)
Comment (Lan2u) [reply boa/src/syntax/parser/statement/declaration/hoistable/async_function_decl/mod.rs:53]: Yeah makes sense, I'll modify peek_expect_no_lineterminator

File: boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Razican) [reply boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs:67]: As far as I understand, there are 3 options, either, we pass the same value with `?`, we pass `false` with `~` or we pass `true` with `+`. Maybe @jasonwilliams or @HalidOdat can confirm this.

[the remark this answers is not in the document]
Comment (Lan2u) [reply boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs:67]: To be honest even after reading the spec (https://tc39.es/ecma262/#sec-grammar-notation) I'm still unsure - I took their usage of 'not' to mean logical inverse but perhaps they meant logical false

File: boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+use std::io::Read;
+
+/// Async Function expression parsing.
+///
+/// More information:
+///  - [MDN documentation][mdn]
+///  - [ECMAScript specification][spec]
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/async_function
+/// [spec]: https://www.ecma-international.org/ecma-262/11.0/index.html#prod-AsyncFunctionExpression
```
[the remark this answers is not in the document]
Comment (Lan2u) [reply boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs:26]: Fixed

File: boa/src/syntax/parser/expression/primary/object_initializer/mod.rs [hunk 14]
```diff
@@ -141,6 +141,14 @@ where
             return Ok(node::PropertyDefinition::property(prop_name, val));
         }
 
+        // TODO GeneratorMethod
```
[the remark this answers is not in the document]
Comment (Lan2u) [reply boa/src/syntax/parser/expression/primary/object_initializer/mod.rs:144]: Good idea, done

File: boa/src/syntax/parser/expression/primary/mod.rs [hunk 15]
```diff
@@ -77,6 +78,9 @@ where
             TokenKind::Keyword(Keyword::Function) => {
                 FunctionExpression.parse(cursor).map(Node::from)
             }
+            TokenKind::Keyword(Keyword::Async) => AsyncFunctionExpression::new(self.allow_yield)
```
[the remark this answers is not in the document]
Comment (Lan2u) [reply boa/src/syntax/parser/expression/primary/mod.rs:81]: Not in this PR, I'm planning to do async methods/arrow functions next

File: boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Razican) [reply boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs:68]: Looking at this section of the spec, if it's `+`, then, we only take into account when that parameter is set (`true` in Boa), when it's `?`, then, it will depend on that parameter being `true` or `false` in the current production, so we just pass it, and if it's `~`, then, we only take into account when that parameter is not set, so `false` in Boa:
Screenshot from 2020-10-12 13-26-21

[the remark this answers is not in the document]
Comment (Lan2u) [reply boa/src/syntax/parser/expression/primary/async_function_expression/mod.rs:68]: Hmm ok

**Review by HalidOdat — APPROVED** [review]
